crates/topos-tce-broadcast/src/task_manager/mod.rs

```diff
@@ -66,35 +70,47 @@ impl TaskManager {
         message_signer: Arc<MessageSigner>,
         validator_store: Arc<ValidatorStore>,
         broadcast_sender: broadcast::Sender<CertificateDeliveredWithPositions>,
-    ) -> (Self, mpsc::Receiver<()>) {
-        let (shutdown_sender, shutdown_receiver) = mpsc::channel(1);
-
-        (
-            Self {
-                message_receiver,
-                task_completion_sender,
-                subscriptions,
-                event_sender,
-                tasks: HashMap::new(),
-                running_tasks: FuturesUnordered::new(),
-                buffered_messages: Default::default(),
-                validator_id,
-                message_signer,
-                thresholds,
-                shutdown_sender,
-                validator_store,
-                broadcast_sender,
-                precedence: HashMap::new(),
-            },
-            shutdown_receiver,
-        )
+    ) -> Self {
+        Self {
+            message_receiver,
+            task_completion_sender,
+            subscriptions,
+            event_sender,
+            tasks: HashMap::new(),
+            running_tasks: FuturesUnordered::new(),
+            buffered_messages: Default::default(),
+            validator_id,
+            message_signer,
+            thresholds,
+            validator_store,
+            broadcast_sender,
+            latest_pending_id: 0,
+        }
     }
 
-    pub async fn run(mut self, mut shutdown_receiver: mpsc::Receiver<()>) {
+    pub async fn run(mut self, mut shutdown_receiver: CancellationToken) {
+        let mut interval = tokio::time::interval(Duration::from_secs(1));
+
         loop {
             tokio::select! {
                 biased;
 
+                _ = interval.tick() => {
+                    debug!("Checking for next pending_certificates");
+                    match self.validator_store.get_next_pending_certificates(&self.latest_pending_id, 1000) {
```

Comment:
The amount (`1000`) should be a `const` somewhere, or be able to passed down. Also, where does the number come from? Doesn't the `storage` know how many are in pending, and we just pass the latest `pending_id` down?

Reply:
The number `1000` is a random-picked value, and having a `limit` on the method allow any other app to require only the amount of certificate in pending that they want.